ykrt/src/mt.rs

```diff
@@ -283,25 +295,40 @@ impl MTThread {
                     };
 
                     if stop_tracer {
-                        if pack.compare_and_swap(lp, PHASE_COMPILED, Ordering::Release) == lp {
-                            let _sir_trace = Rc::get_mut(&mut self.inner)
-                                .unwrap()
-                                .tracer
-                                .take()
-                                .unwrap()
-                                .0
-                                .stop_tracing();
-                            // FIXME build TIR and compile. Eventually in a background thread?
+                        let sir_trace = Rc::get_mut(&mut self.inner)
+                            .unwrap()
+                            .tracer
+                            .take()
+                            .unwrap()
+                            .0
+                            .stop_tracing()
+                            .unwrap();
+                        let tir_trace = TirTrace::new(&*SIR, &*sir_trace).unwrap();
+                        let ct = TraceCompiler::<I>::compile(tir_trace);
+                        let ptr = Box::into_raw(Box::new(ct)) as usize; //ct.ptr() as usize;
+
+                        let new_pack = PHASE_COMPILED | (ptr << PHASE_NUM_BITS);
+                        if pack.compare_and_swap(lp, new_pack, Ordering::Release) == lp {
                             break;
                         }
                     } else {
                         break;
                     }
                 }
-                PHASE_COMPILED => break, // FIXME call compiled trace, but don't call step_fn().
+                PHASE_COMPILED => {
+                    if pack.load(Ordering::Acquire) == lp {
```

Comment:
Why do we need to reload `pack` from memory? Can't we just operate on `lp`?

Reply:
Good point.

Reply:
You're right. The atomic shouldn't have changed between this and the previous `load`.

Reply:
Fixed in bc9291b.